rcl_yaml_param_parser/src/parse.c

```diff
@@ -404,6 +409,147 @@ rcutils_ret_t parse_value(
   return ret;
 }
 
+///
+/// Check a name space whether it is valid
+///
+static rcutils_ret_t
+__validate_namespace(const char * namespace)
+{
+  int validation_result = 0;
+  rmw_ret_t ret;
+  ret = rmw_validate_namespace(namespace, &validation_result, NULL);
+  if (RMW_RET_OK != ret) {
+    RCUTILS_SET_ERROR_MSG(rmw_get_error_string().str);
+    return RCUTILS_RET_ERROR;
+  }
+  if (RMW_NAMESPACE_VALID != validation_result) {
+    RCUTILS_SET_ERROR_MSG(rmw_namespace_validation_result_string(validation_result));
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+///
+/// Check a node name whether it is valid
+///
+static rcutils_ret_t
+__validate_nodename(const char * name)
+{
+  int validation_result = 0;
+  rmw_ret_t ret;
+  ret = rmw_validate_node_name(name, &validation_result, NULL);
+  if (RMW_RET_OK != ret) {
+    RCUTILS_SET_ERROR_MSG(rmw_get_error_string().str);
+    return RCUTILS_RET_ERROR;
+  }
+  if (RMW_NODE_NAME_VALID != validation_result) {
+    RCUTILS_SET_ERROR_MSG(rmw_node_name_validation_result_string(validation_result));
+    return RCUTILS_RET_ERROR;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+///
+/// Check a name (namespace/node_name) whether it is valid
+///
+static rcutils_ret_t
+__validate_name(const char * name, rcutils_allocator_t allocator)
+{
+  // special rules
+  if (0 == strcmp(name, "/**") || 0 == strcmp(name, "/*")) {
+    return RCUTILS_RET_OK;
+  }
+
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  char * separator_pos = rindex(name, '/');
+  char * node_name = NULL;
+  char * absolute_namespace = NULL;
+  if (NULL == separator_pos) {
+    node_name = rcutils_strdup(name, allocator);
+    if (NULL == node_name) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+  } else {
+    // substring namespace including the last '/'
+    char * namespace = rcutils_strndup(name, separator_pos - name + 1, allocator);
+    if (NULL == namespace) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+    if (namespace[0] != '/') {
+      absolute_namespace = rcutils_format_string(allocator, "/%s", namespace);
+      allocator.deallocate(namespace, allocator.state);
+      if (NULL == absolute_namespace) {
+        ret = RCUTILS_RET_BAD_ALLOC;
+        goto clean;
+      }
+    } else {
+      absolute_namespace = namespace;
+    }
+
+    node_name = rcutils_strdup(separator_pos + 1, allocator);
+    if (NULL == node_name) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+  }
+
+  if (absolute_namespace) {
+    size_t i = 0;
+    separator_pos = index(absolute_namespace + i + 1, '/');
+    if (NULL == separator_pos) {
+      ret = __validate_namespace(absolute_namespace);
+      if (RCUTILS_RET_OK != ret) {
+        goto clean;
+      }
+    } else {
+      do {
```

Comment:
I think we can refactor the following logic so it is easier to read and maintain. Instead of making repeated allocations in the loop, what if we just did a find-and-replace of all `*` characters in the string with a random valid character (e.g. `*` -> `W`)? Then pass the result to `__validate_namespace`. It seems like it would be a lot simpler to me. The only caveat is that we risk running over the limit `RMW_NAMESPACE_MAX_LENGTH`, but I'm not sure that is a case we should worry about handling.

Reply:
Actually, I thought that before. If it's partial matches for wildcards, such as '/foo/**part/node', this policy failed.

Reply:
Ah, yeah that makes sense. Nevermind then.